### ISSUE TITLE
Some matmul tests are failing on RTX4090

### ISSUE THREAD

**Issue by zasdfgbnm** [NVIDIA/Fuser#601]
```C++
[ RUN      ] NVFuserTest.FusionAmpereMatmulSmemEpilogueCast_CUDA
unknown file: Failure
C++ exception with description "getStaticSmemSize() + dynamic_smem_size < device_smem_limit_ INTERNAL ASSERT FAILED at "/home/gaoxiang/Fuser/csrc/executor.cpp":1516, please report a bug to PyTorch. The total shared memory allocation is larger than available memory. Dynamic size: 131072. Static size: 0. Required total size: 131072. Device limit size: 101376
Exception raised from validateDynamicSmemSize at /home/gaoxiang/Fuser/csrc/executor.cpp:1516 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7fc23e5a8477 in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7fc23e56489b in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::string const&) + 0x3f (0x7fc23e5a61cf in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #3: nvfuser::FusionExecutor::validateDynamicSmemSize(long) + 0xda (0x7fc23e8b7b0a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #4: nvfuser::FusionExecutor::ensureAvailableDynamicSmemSize(long) + 0x4a (0x7fc23e8b7b6a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #5: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams) + 0xdf2 (0x7fc23e8c5ff2 in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #6: <unknown function> + 0x1ee0df (0x55d9662db0df in ./build/nvfuser_tests)
frame #7: <unknown function> + 0x572d7f (0x55d96665fd7f in ./build/nvfuser_tests)
frame #8: <unknown function> + 0x666917 (0x55d966753917 in ./build/nvfuser_tests)
frame #9: <unknown function> + 0x658fed (0x55d966745fed in ./build/nvfuser_tests)
frame #10: <unknown function> + 0x6591c5 (0x55d9667461c5 in ./build/nvfuser_tests)
frame #11: <unknown function> + 0x6593a0 (0x55d9667463a0 in ./build/nvfuser_tests)
frame #12: <unknown function> + 0x65cb7d (0x55d966749b7d in ./build/nvfuser_tests)
frame #13: <unknown function> + 0x666ea7 (0x55d966753ea7 in ./build/nvfuser_tests)
frame #14: <unknown function> + 0x6595d2 (0x55d9667465d2 in ./build/nvfuser_tests)
frame #15: <unknown function> + 0x189019 (0x55d966276019 in ./build/nvfuser_tests)
frame #16: <unknown function> + 0x23850 (0x7fc1d9239850 in /usr/lib/libc.so.6)
frame #17: __libc_start_main + 0x8a (0x7fc1d923990a in /usr/lib/libc.so.6)
frame #18: <unknown function> + 0x1c09a5 (0x55d9662ad9a5 in ./build/nvfuser_tests)
" thrown in the test body.
[  FAILED  ] NVFuserTest.FusionAmpereMatmulSmemEpilogueCast_CUDA (1307 ms)
```

```C++
[ RUN      ] NVFuserTest.FusionAmpereMatmulSmemEpilogueRelu_CUDA
unknown file: Failure
C++ exception with description "getStaticSmemSize() + dynamic_smem_size < device_smem_limit_ INTERNAL ASSERT FAILED at "/home/gaoxiang/Fuser/csrc/executor.cpp":1516, please report a bug to PyTorch. The total shared memory allocation is larger than available memory. Dynamic size: 131072. Static size: 0. Required total size: 131072. Device limit size: 101376
Exception raised from validateDynamicSmemSize at /home/gaoxiang/Fuser/csrc/executor.cpp:1516 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7fc23e5a8477 in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7fc23e56489b in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::string const&) + 0x3f (0x7fc23e5a61cf in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #3: nvfuser::FusionExecutor::validateDynamicSmemSize(long) + 0xda (0x7fc23e8b7b0a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #4: nvfuser::FusionExecutor::ensureAvailableDynamicSmemSize(long) + 0x4a (0x7fc23e8b7b6a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #5: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams) + 0xdf2 (0x7fc23e8c5ff2 in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #6: <unknown function> + 0x1ee0df (0x55d9662db0df in ./build/nvfuser_tests)
frame #7: <unknown function> + 0x573da7 (0x55d966660da7 in ./build/nvfuser_tests)
frame #8: <unknown function> + 0x666917 (0x55d966753917 in ./build/nvfuser_tests)
frame #9: <unknown function> + 0x658fed (0x55d966745fed in ./build/nvfuser_tests)
frame #10: <unknown function> + 0x6591c5 (0x55d9667461c5 in ./build/nvfuser_tests)
frame #11: <unknown function> + 0x6593a0 (0x55d9667463a0 in ./build/nvfuser_tests)
frame #12: <unknown function> + 0x65cb7d (0x55d966749b7d in ./build/nvfuser_tests)
frame #13: <unknown function> + 0x666ea7 (0x55d966753ea7 in ./build/nvfuser_tests)
frame #14: <unknown function> + 0x6595d2 (0x55d9667465d2 in ./build/nvfuser_tests)
frame #15: <unknown function> + 0x189019 (0x55d966276019 in ./build/nvfuser_tests)
frame #16: <unknown function> + 0x23850 (0x7fc1d9239850 in /usr/lib/libc.so.6)
frame #17: __libc_start_main + 0x8a (0x7fc1d923990a in /usr/lib/libc.so.6)
frame #18: <unknown function> + 0x1c09a5 (0x55d9662ad9a5 in ./build/nvfuser_tests)
" thrown in the test body.
[  FAILED  ] NVFuserTest.FusionAmpereMatmulSmemEpilogueRelu_CUDA (1291 ms)
```

```C++
[ RUN      ] MatmulSASSTest.AmpereModifiers_CUDA
unknown file: Failure
C++ exception with description "getStaticSmemSize() + dynamic_smem_size < device_smem_limit_ INTERNAL ASSERT FAILED at "/home/gaoxiang/Fuser/csrc/executor.cpp":1516, please report a bug to PyTorch. The total shared memory allocation is larger than available memory. Dynamic size: 131072. Static size: 0. Required total size: 131072. Device limit size: 101376
Exception raised from validateDynamicSmemSize at /home/gaoxiang/Fuser/csrc/executor.cpp:1516 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7fc23e5a8477 in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7fc23e56489b in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::string const&) + 0x3f (0x7fc23e5a61cf in /home/gaoxiang/.local/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #3: nvfuser::FusionExecutor::validateDynamicSmemSize(long) + 0xda (0x7fc23e8b7b0a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #4: nvfuser::FusionExecutor::ensureAvailableDynamicSmemSize(long) + 0x4a (0x7fc23e8b7b6a in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #5: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams) + 0xdf2 (0x7fc23e8c5ff2 in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #6: <unknown function> + 0x57e13a (0x55d96666b13a in ./build/nvfuser_tests)
frame #7: <unknown function> + 0x57ec00 (0x55d96666bc00 in ./build/nvfuser_tests)
frame #8: <unknown function> + 0x666917 (0x55d966753917 in ./build/nvfuser_tests)
frame #9: <unknown function> + 0x658fed (0x55d966745fed in ./build/nvfuser_tests)
frame #10: <unknown function> + 0x6591c5 (0x55d9667461c5 in ./build/nvfuser_tests)
frame #11: <unknown function> + 0x6593a0 (0x55d9667463a0 in ./build/nvfuser_tests)
frame #12: <unknown function> + 0x65cb7d (0x55d966749b7d in ./build/nvfuser_tests)
frame #13: <unknown function> + 0x666ea7 (0x55d966753ea7 in ./build/nvfuser_tests)
frame #14: <unknown function> + 0x6595d2 (0x55d9667465d2 in ./build/nvfuser_tests)
frame #15: <unknown function> + 0x189019 (0x55d966276019 in ./build/nvfuser_tests)
frame #16: <unknown function> + 0x23850 (0x7fc1d9239850 in /usr/lib/libc.so.6)
frame #17: __libc_start_main + 0x8a (0x7fc1d923990a in /usr/lib/libc.so.6)
frame #18: <unknown function> + 0x1c09a5 (0x55d9662ad9a5 in ./build/nvfuser_tests)
" thrown in the test body.
[  FAILED  ] MatmulSASSTest.AmpereModifiers_CUDA (1099 ms)
```

**Comment by liqiangxl** [comment]
I also noticed several other failed cases on the same node due to validation errors. I also tried an earlier commit `864bcd8` before shared memory epilogue, same results.

[  FAILED  ] 7 tests, listed below:
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueOutputCast_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueAlpha_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaOutputCast_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueRelu_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueGelu_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaBeta_CUDA
[  FAILED  ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaBetaGeluOutputCast_CUDA

**Comment by zasdfgbnm** [comment]
> > I also noticed several other failed cases on `RTX 3080 @ nvdl-bsg-amd02` due to validation errors. I also tried an earlier commit `864bcd8` before shared memory epilogue, same results.
> > [ FAILED ] 7 tests, listed below: [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueOutputCast_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueAlpha_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaOutputCast_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueRelu_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueGelu_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaBeta_CUDA [ FAILED ] NVFuserTest.FusionMatmulSchedulerEpilogueAlphaBetaGeluOutputCast_CUDA
> 
> Do we need to increase the tolerance for these cases?

Please do :)

**Comment by liqiangxl** [comment]
Realized it is more reasonable to calculate the reference in float. Fixed in #606 and #607
